### ISSUE TITLE
If player receives a formspec while in the death screen, the death screen disappears, breaking the game

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#11523]
##### Minetest version
```
5.5.0
```

##### Summary
If you receive a formspec with `minetest.show_formspec` while you're in the death screen, the death screen is overwritten by the new formspec. You can close the new formspec normally and you now can walk, while still being dead (0 HP), around the world but you can't interact with it and you can't respawn. It's a very broken state of gameplay that causes lots of bugs. The only way to get out of this "zombie state" is to rejoin.

##### Steps to reproduce
1. Get `luacmd` mod
2. Start any game
3. `/lua me:set_hp(0); minetest.after(3, function() minetest.show_formspec("singleplayer", "title", "size[5,5]") end)`
4. Do nothing for 3 seconds

Expected result:
You should die and the death screen is not overwritten, nor will you end up in a weird or broken player state.

Actual result:
You will die, but after 3 seconds, the death screen is replaced by a simple dummy formspec (see above). You can close it and walk around the world with 0 HP. You can't respawn because the respawn button, which is the only way to respawn, is gone for good.

##### Possible solutions
A) While the death screen is open, reject any incoming formspecs so that the death screen and the respawn button does not disappear.
B) Open the formspec as usual, but when the player closes it, the Death Screen immediately appears again. Basically, the idea is to instantly open the death screen when HP=0 and no "normal" formspec is open
C) Refactor the Death Screen completely. Instead of using a normal formspec
C) Something else?

I personally prefer solution A because it's the simplest but it will need proper documentation and a warning in `lua_api.txt`.

##### Related issues

* #7879
* #2945

**Comment by sfan5** [comment]
This is effectively the same bug as #7879, just the cause is different.

**Comment by Wuzzy2** [comment]
But we don't know the cause of both issues so far. If there are two causes for the bug, that's worth two issues.
Because in that case, if you fix one of the two issues, the other one still remains.

**Comment by sfan5** [comment]
True. I was thinking that 
>B) Open the formspec as usual, but when the player closes it, the Death Screen immediately appears again. Basically, the idea is to instantly open the death screen when HP=0 and no "normal" formspec is open

would fix both of the bugs at once.

**Comment by ancientmarinerdev** [comment]
> True. I was thinking that
> 
> > B) Open the formspec as usual, but when the player closes it, the Death Screen immediately appears again. Basically, the idea is to instantly open the death screen when HP=0 and no "normal" formspec is open
> 
> would fix both of the bugs at once.

Related MineClone2 issue:

https://git.minetest.land/MineClone2/MineClone2/issues/1428

and here:

https://forum.minetest.net/viewtopic.php?p=430578#p430578

Some within our issue tracker suspect it can happen when you already have a formspec open. Maybe sporadically (every 30s, when dead, bring up the formspec to handle cases where user logs off, and back in so they don't get stuck in a state they cannot get out of it. So yeah, I think it's the same bug, but a few manifestations to navigate around.

**Comment by unersame342** [comment]
That's true that's complicated if we want to include customized HUD into the world at connection for example..
happen when you're disconnected when you're still on the respawn page and haven't clicked on the button and try to reconnect, you are lying on the ground and can walk but not die or respawn .. Strange..

**Comment by Wuzzy2** [comment]
I have thought of a better solution now:

1) The client keeps receiving and accepting formspecs as usual
2) If the client receives a formspec when neither respawn or Esc screen are up: Display formspec
3) If the client receives a formspec if the respawn or Esc screen are up, do NOT display the formspec, but DO keep it in memory. Do NOT close or replace the respawn or Esc screen
4) When the player closes the Esc or respawn screen while no formspec is in memory: Close it
5) When the player closes the Esc or respawn screen while a formspec is in memory: Immediately replace it with the formspec from memory
6) When the player closes a normal formspec: Delete the formspec from memory

This essentially will turn the Esc/respawn screen into a modal window. The normal formspec is still there, it just is hidden and can't be interacted with while the respawn or Esc screen are up. The *current* behavior is that the normal formspec completely destroys the respawn/Esc screen, which is a problem.

This should also make it easy for games because they can send formspecs at any time and don't have to perform any special checks or workarounds.

(Oh, I forgot to mention: Any sent formspec can also happily destroy the Esc screen as well. Which isn't as serious of a bug, but games should never be able to that, it's bad UX)

**Comment by sfan5** [comment]
short discussion: https://irc.luanti.org/luanti-dev/2026-01-25#i_6311307
